pkg/alexa/gen/skill.go

```diff
@@ -6,415 +6,283 @@ import (
-	s.category = category
+func (s *SkillBuilder) WithLocaleRegistry(registry l10n.LocaleRegistry) *SkillBuilder {
+	s.registry = registry
 	return s
 }
 
-func (s *SkillBuilder) SetCategory(category alexa.Category) {
+func (s *SkillBuilder) WithCategory(category alexa.Category) *SkillBuilder {
 	s.category = category
+	return s
 }
 
-func (s *SkillBuilder) SetTestingInstructions(instructions string) {
-	s.testingInstructions = instructions
-}
-
-func (s *SkillBuilder) SetDefaultLocale(locale string) {
-	s.defaultLocale = locale
+func (s *SkillBuilder) WithCountries(countries []string) *SkillBuilder {
+	s.countries = countries
+	return s
 }
 
-func (s *SkillBuilder) SetModelDelegation(delegation alexa.DialogDelegation) {
-	s.modelDelegation = delegation
+func (s *SkillBuilder) WithTestingInstructions(instructions string) *SkillBuilder {
+	s.skillInstructions = instructions
+	return s
 }
 
-func (s *SkillBuilder) AddLocale(l string, trans l10n.LocaleInstance) {
-	if len(s.locales) == 0 {
-		// ensure that a default is set
-		if s.defaultLocale == "" {
-			s.defaultLocale = l
-		}
+func (s *SkillBuilder) WithLocaleTestingInstructions(locale string, instructions string) *SkillBuilder {
+	loc, err := s.registry.Resolve(locale)
+	if err != nil {
+		return s
 	}
-	s.locales[l] = LocaleDef{Translations: trans}
+	loc.Set(s.skillInstructions, []string{instructions})
+	return s
 }
 
-func (s *SkillBuilder) AddIntent(intent Intent) {
-	s.intents = append(s.intents, intent)
+func (s *SkillBuilder) WithPrivacyFlag(flag string, value bool) *SkillBuilder {
+	s.privacyFlags[flag] = value
+	return s
 }
 
-// AddIntentString creates an Intent from the string, adds it and returns a reference.
-func (s *SkillBuilder) AddIntentString(intent string) *Intent {
-	i := NewIntent(intent)
-	s.intents = append(s.intents, i)
-	return &i
-}
+func (s *SkillBuilder) AddLocale(locale string) *SkillLocaleBuilder {
+	s.registry.Register(l10n.NewLocale(locale))
 
-func (s *SkillBuilder) AddType(t Type) {
-	// TODO: validate rules!
-	s.types = append(s.types, t)
+	lb := NewSkillLocaleBuilder(locale).
+		WithLocaleRegistry(s.registry)
+	s.locales[locale] = lb
+	return lb
 }
 
-func (s *SkillBuilder) AddTypeString(t string) {
-	s.types = append(s.types, Type(t))
-}
-
-func (s *SkillBuilder) AddCountry(c alexa.Country) {
-	s.countries = append(s.countries, c)
+func (s *SkillBuilder) AddCountry(country string) *SkillBuilder {
+	s.countries = append(s.countries, country)
+	return s
 }
 
-func (s *SkillBuilder) AddCountries(cs []alexa.Country) {
+func (s *SkillBuilder) AddCountries(cs []string) *SkillBuilder {
 	for _, c := range cs {
 		s.countries = append(s.countries, c)
 	}
+	return s
 }
 
 // Build builds an alexa.Skill object.
-// TODO: return errors!
 func (s *SkillBuilder) Build() (*alexa.Skill, error) {
+	if s.registry == nil || len(s.registry.GetLocales()) == 0 {
+		return nil, fmt.Errorf("No locales registered to build")
+	}
+	//
+	if s.locales == nil || len(s.locales) == 0 {
```

Comment:
This is an unexpected turn. I didnt give you any, so you take what I have in the registry. Seems magical and unexpected.

Reply:
That is where the two approaches (provide locale translations manually vs. use registry) start to interfere, you should only use one or know what you're doing :)